#67 Create volume restful api docs (#69)
oharastream/ohara-docs: content/en/docs/master/rest-api/volumes.md

---
title: Volume
linktitle: Volume API
toc: true
type: docs
data: "2020-11-27T00:00:00+01:00"
draft: false
menu:
  master:
    parent: REST APIs
    weight: 200
---

Ohara volume feature is persisting data by zookeeper container and broker container.

Note: Confirm your user name is ohara and UUID is 1000 in the running cluster environment.

The properties which can be set by user are shown below.

1. name (**string**) --- volume name. The legal character is lowercase alphanumerics characters and number.

2. group (**string**) --- volume group. The legal character is lowcase alphanumerics characters and number.

3. path (**string**) --- The data folder path in the node.

> 4. nodeNames (**string**) --- Setting the node to create ohara volume for the zookeeper and broker container use.

Reviewer: IIRC, the type of nodeNames should be `Set[String]` instead of `String`?

Author: 寫錯了，我來修改。

We will use the default value as the query parameter "?group=" if you don't specify it.

* Example Resposne
    ```
    202 Accepted
    ```

{{% alert hint %}}
You should use [Get volume info]({{< relref "#get" >}}) to show volume up-to-date status.

## get a volume info {#get}

*GET /v0/volumes/$name?group=$group*

Get volume info by name and group. This API could fetch all volume info.
We will use the default value as the query parameter "?group=" if you don't
specify it.

* Example Response
    ```json
    {
      "group": "default",
      "lastModified": 1606457846012,
      "name": "volume1",
      "nodeNames": [
        "k8s-master"
      ],
      "path": "/tmp/workspace1/bk12345",
      "state": "RUNNING",
      "tags": {}
     }
     ```

## add a volume to a running volumes

*PUT /v0/volumes/$name/$nodeName?group=$group*

Add a new volume for other node. We will use the default value as the query parameter "?group=" if you don't specify it.

* Example Resposne
    ```
    202 Accepted
    ```

## stop a volume info

Stop a running volume and remove the volume data folder (dangerous). It is disallowed to stop a volume cluster used by a running.

> *PUT /v0/volumes/$name/stop?group=$group[&force=true]*

Reviewer: We could force stop a volume?

Author: K8S 的 implement 有實作 volume force 的部份

Reviewer: Got it, thanks 👍

---
title: Volume
linktitle: Volume API
toc: true
type: docs
data: "2020-11-27T00:00:00+01:00"
draft: false
menu:
  master:
    parent: REST APIs
    weight: 200
---

Ohara volume feature is persisting data by zookeeper container and broker container.

Note: Confirm your user name is ohara and UUID is 1000 in the running cluster environment.

The properties which can be set by user are shown below.

> 1. name (**String**) --- volume name. The legal character is lowercase alphanumerics characters and number.

Reviewer 2: ```suggestion
1. name (**String**) --- The legal character is lowercase alphanumerics characters and number.
```

Author: ok, 我來修改。

> 2. group (**String**) --- volume group. The legal character is lowcase alphanumerics characters and number.

Reviewer 2: ```suggestion
2. group (**String**) --- The legal character is lowercase alphanumerics characters and number.
```

Author: ok, 我來修改。

3. path (**String**) --- The data folder path in the node.

> 4. nodeNames (**Set[String]**) --- Setting the node to create ohara volume for the zookeeper and broker container use.

Reviewer 2: ```suggestion
4. nodeNames (**Set[String]**) --- The nodes to create ohara volume for zookeeper and broker containers.
```

Author: 了解，我來修改。

> 5. tags (**Object**) --- the user defined parameters.  

Reviewer 2: ```suggestion
5. tags (**Object**) --- User-defined parameters.  
```

Author: OK, 我來修改。thank you!

## start a ohara volume

*PUT /v0/volumes/$name/start?group=$group*

We will use the default value as the query parameter "?group=" if you don't specify it.

* Example Resposne
    ```
    202 Accepted
    ```

{{% alert hint %}}
You should use [Get volume info]({{< relref "#get" >}}) to show volume up-to-date status.

## get a volume info {#get}

*GET /v0/volumes/$name?group=$group*

Get volume info by name and group. This API could fetch all volume info.
We will use the default value as the query parameter "?group=" if you don't
specify it.

* Example Response
    ```json
    {
      "group": "default",
      "lastModified": 1606457846012,
      "name": "volume1",
      "nodeNames": [
        "k8s-master"
      ],
      "path": "/tmp/workspace1/bk12345",
      "state": "RUNNING",
      "tags": {}
     }
     ```

## add a volume to a running volumes

*PUT /v0/volumes/$name/$nodeName?group=$group*

Add a new volume for other node. We will use the default value as the query parameter "?group=" if you don't specify it.

* Example Resposne
    ```
    202 Accepted
    ```

## update volume properties setting

*PUT /v0/volumes/$name?group=$group*

> If the required volume (group, name) was not exists, we will try to use this request as POST.

Reviewer 2: ```suggestion
If the required volume (group, name) doesn't exist, we will try to use this request as POST.
```

Author: ok, 我來修改。

* Example Request
    ```json
    {
      "nodeNames": [
        "node00", "node01"
      ],
      "path": "/tmp/workspace2"
    }
    ```
* Example Response
    ```json
    {
      "group": "default",
      "lastModified": 1606465299352, 
      "name": $name,
      "nodeNames": ["node00", "node01"],
      "path": "/tmp/workspace2",
      "tags": {}
    }
    ```
## stop a volume info

> Stop a running volume and remove the volume data folder (dangerous). It is disallowed to stop a volume cluster used by a running.

Reviewer 2: This seems like an unfinshed sentence? by a running __?

Reviewer 2: I'm guessing _service_ or _cluster_?

Author: zookeeper or broker container.
